lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumOutputScalars[i], 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetInputToOutputTable(i), ViewIDState.InputToOutputTable[i],
+        PSV.GetViewIDOutputMask(i), ViewIDState.ViewIDOutputMask[i],
+        NumInputVectors, NumOutputVectors, PSV1->UsesViewID);
+  }
+
+  // NumInputScalars as input, NumPCOrPrimScalars as output.
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Hull) ||
+      PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Mesh)) {
+    uint32_t NumInputVectors = llvm::RoundUpToAlignment(NumInputScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetInputToPCOutputTable(), ViewIDState.InputToPCOutputTable,
+        PSV.GetViewIDPCOutputMask(), ViewIDState.ViewIDPCOutputMask,
+        NumInputVectors, NumOutputVectors, PSV1->UsesViewID);
+  }
+
+  // NumPCOrPrimScalars as input, NumOutputScalars[0] as output.
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Domain)) {
+    uint32_t NumInputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumOutputScalars[0], 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetPCInputToOutputTable(), ViewIDState.PCInputToOutputTable,
+        PSVComponentMask(), ViewIDState.ViewIDOutputMask[0], NumInputVectors,
+        NumOutputVectors, false);
+  }
+  if (Mismatched) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    PSV.PrintViewIDState(OS);
+    OS.flush();
+    std::string Str1;
+    raw_string_ostream OS1(Str1);
+    ViewIDState.Print(OS1, static_cast<PSVShaderKind>(PSV1->ShaderStage),
+                      PSV1->UsesViewID);
+    OS1.flush();
+    EmitMismatchError("ViewIDState", Str, Str1);
+  }
+}
+
+void PSVContentVerifier::VerifySignatures(unsigned ValMajor,
+                                          unsigned ValMinor) {
+  bool i1ToUnknownCompat = DXIL::CompareVersions(ValMajor, ValMinor, 1, 5) < 0;
+  // Verify input signature
+  VerifySignature(DM.GetInputSignature(), PSV.GetInputElement0(0),
+                  PSV.GetSigInputElements(), "SigInput", i1ToUnknownCompat);
+  // Verify output signature
+  VerifySignature(DM.GetOutputSignature(), PSV.GetOutputElement0(0),
+                  PSV.GetSigOutputElements(), "SigOutput", i1ToUnknownCompat);
+  // Verify patch constant signature
+  VerifySignature(DM.GetPatchConstOrPrimSignature(),
+                  PSV.GetPatchConstOrPrimElement0(0),
+                  PSV.GetSigPatchConstOrPrimElements(),
+                  "SigPatchConstantOrPrim", i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignature(const DxilSignature &Sig,
+                                         PSVSignatureElement0 *Base,
```

Comment:
Technically, with the PSV0 versioning scheme, this pattern that assumes this will always be an array of `PSVSignatureElement0` is not safe.  `DxilPipelineStateValidation::m_uPSVSignatureElementSize` must be used to index into the elements in the PSV0 data.

Furthermore, that stride (and all strides) should be checked for validity based on the version.  Loading will only check that `sizeof(PSVSignatureElement0) <= m_uPSVSignatureElementSize`.

The loader can handle a newer version of the format, but the validator needs to make sure the format is constrained to what it knows is valid.  This means the validator should verify that `sizeof(PSVSignatureElement0) == m_uPSVSignatureElementSize` for all versions of PSV0 defined so far.

I see that the record strides are not part of the public interface of the `DxilPipelineStateValidation` class though, so these may need exposure to enable both stride validation, and use of the `GetRecord<>()` accessor based on base pointer, stride, and element count.

Reply:
Switched to GetRecord instead of Base + i.